File: tests/java/verifiers/src/main/java/io/vectorized/compaction/tx/TxWorkload.java
package io.vectorized.compaction.tx;
import com.google.gson.Gson;
import io.vectorized.compaction.App;
import io.vectorized.compaction.GatedWorkload;
import java.lang.Thread;
import java.time.Duration;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Properties;
import java.util.Queue;
import java.util.Random;
import java.util.UUID;
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.consumer.ConsumerRecords;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.Producer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.TopicPartition;

public class TxWorkload extends GatedWorkload {
  public static class InitBody {
    public String brokers;
    public String topic;
    public int partitions;
    public int key_set_cardinality;
  }

  static class WriteInfo {
    public long id;
    public String key;
    public int partition;
    public long offset;
  }

  static class LatestValue {
    public WriteInfo confirmed;
    public Queue<WriteInfo> attempts;

    public LatestValue() {
      this.confirmed = null;
      this.attempts = new LinkedList<>();
    }
  }

  static class Value {
    public long id;
    public long offset;

    public Value(long id, long offset) {
      this.id = id;
      this.offset = offset;
    }
  }

  static class Partition {
    public int id;
    public HashMap<Long, WriteInfo> writes;       // hashed by WriteInfo.id
    public HashMap<String, LatestValue> latestKV; // hashed by key

    public long countWritten = 0;
    public long countRead = 0;

    public long writtenOffset = -1;
    public long endOffset = -1;
    public long readOffset = -1;
    public long readPosition = -1;
    public boolean consumed = false;

    public Partition(int id) {
      this.id = id;
      this.writes = new HashMap<>();
      this.latestKV = new HashMap<>();
    }
  }

  private static final int BATCH_SIZE = 10;

  HashMap<Integer, Partition> partitions;
  private volatile InitBody args;

  public TxWorkload(String params) {
    Gson gson = new Gson();
    this.args = gson.fromJson(params, InitBody.class);
  }

  @Override
  public void startProducer() throws Exception {
    super.startProducer();

    System.out.println(
        "#logical_time\tseconds_since_start\tpartition\tlog_type\tdetails");

    log(-1, "start\tproducer");

    this.write_threads = new ArrayList<>();
    this.partitions = new HashMap<>();

    for (int partition = 0; partition < args.partitions; partition++) {
      this.partitions.put(partition, new Partition(partition));

      final int p = partition;
      this.write_threads.add(new Thread(() -> {
        try {
          writeProcess(p);
        } catch (Exception e) {
          synchronized (this) {
            System.out.println("=== write process error");
            System.out.println(e);
            e.printStackTrace(System.out);
            System.exit(1);
          }
        }
      }));
    }

    for (var th : write_threads) {
      th.start();
    }
  }

  @Override
  public void startConsumer() throws Exception {
    super.startConsumer();

    this.read_threads = new ArrayList<>();

    for (int partition = 0; partition < args.partitions; partition++) {
      final int p = partition;
      this.read_threads.add(new Thread(() -> {
        try {
          readProcess(p);
        } catch (Exception e) {
          synchronized (this) {
            System.out.println("=== read process error");
            System.out.println(e);
            e.printStackTrace(System.out);
            System.exit(1);
          }
        }
      }));
    }

    for (var th : read_threads) {
      th.start();
    }
  }

  private String getKey(Random random) {
    return "key" + random.nextInt(args.key_set_cardinality);
  }

  private void writeProcess(int pid) throws Exception {
    var random = new Random();

    Properties props = new Properties();
    props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, args.brokers);
    props.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, true);
    props.put(
        ProducerConfig.TRANSACTIONAL_ID_CONFIG, UUID.randomUUID().toString());
    props.put(
        ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG,
        "org.apache.kafka.common.serialization.StringSerializer");
    props.put(
        ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG,
        "org.apache.kafka.common.serialization.StringSerializer");
    props.put(ProducerConfig.LINGER_MS_CONFIG, 0);

    var partition = this.partitions.get(pid);
    Producer<String, String> producer = null;

    long id = 0;
    boolean shouldReset = true;
    long lastOffset = -1;

    while (state == State.PRODUCING) {
      if (shouldReset) {
        if (producer != null) {
          try {
            producer.close();
          } catch (Exception e1) {
          }
        }
        producer = null;
        shouldReset = false;
        Thread.sleep(1000);
      }

      if (producer == null) {
        producer = new KafkaProducer<>(props);
        try {
          log(pid, "init");
          producer.initTransactions();
        } catch (Exception e1) {
          synchronized (this) {
            log(pid, "err\tinit");
            System.out.println("=== Error on initTransactions");
            System.out.println(e1);
            e1.printStackTrace(System.out);
          }
          shouldReset = true;
          continue;
        }
      }

      producer.beginTransaction();

      boolean shouldAbort = random.nextInt(3) == 0;

      HashMap<String, WriteInfo> batch = new HashMap<>();

      try {
        for (int i = 0; i < BATCH_SIZE; i++) {
          var op = new WriteInfo();
          op.id = id++;
          op.key = getKey(random);
          op.partition = pid;
          op.offset = -1;

          batch.put(op.key, op);

          if (shouldAbort) {
            log(pid, "a\t" + op.id);
          } else {
            log(pid, "w\t" + op.id);
          }
          String value = "" + op.id + "\t" + (shouldAbort ? "a" : "c") + "\t"
                         + randomString(random, 1024);
          ProducerRecord<String, String> record
              = new ProducerRecord<>(args.topic, pid, op.key, value);
          op.offset = producer.send(record).get().offset();

          if (lastOffset >= op.offset) {
            violation(
                pid, "offsets must me monotonic, written " + op.offset
                         + " while " + lastOffset + " was already known");
            return;
          }
          lastOffset = op.offset;
        }
      } catch (Exception e1) {
        shouldAbort = true;
        synchronized (this) {
          log(pid, "err\tsend");
          System.out.println("=== Error on send");
          System.out.println(e1);
          e1.printStackTrace(System.out);
        }
      }

      if (!shouldAbort) {
        synchronized (this) {
          for (var op : batch.values()) {
            if (!partition.latestKV.containsKey(op.key)) {
              partition.latestKV.put(op.key, new LatestValue());
            }
            var latestValue = partition.latestKV.get(op.key);
            if (!shouldAbort) {
              latestValue.attempts.add(op);
            }
          }
        }
      }

      try {
        if (shouldAbort) {
          log(pid, "abort");
          producer.abortTransaction();
        } else {
          log(pid, "commit");
          producer.commitTransaction();
          synchronized (this) {
            partition.writtenOffset = lastOffset;
            partition.countWritten += BATCH_SIZE;
            for (var op : batch.values()) {
              if (!partition.latestKV.containsKey(op.key)) {
                partition.latestKV.put(op.key, new LatestValue());
              }
              var latestValue = partition.latestKV.get(op.key);
              latestValue.confirmed = op;
              while (latestValue.attempts.size() > 0) {
                var attempt = latestValue.attempts.remove();
                if (attempt.id > op.id) {
                  violation(
                      op.partition,
                      "attempts log jumps over " + op.id + " to " + attempt.id);
                }
                if (attempt.id == op.id) {
                  break;
                }
              }
            }
          }
        }
      } catch (Exception e1) {
        synchronized (this) {
          log(pid, "err");
          System.out.println("=== Error on ending tx");
          System.out.println(e1);
          e1.printStackTrace(System.out);
        }
        shouldReset = true;
        continue;
      }
    }

    producer.close();
  }

  public App.Metrics getMetrics() {
    var metrics = new App.Metrics();
    metrics.min_reads = Long.MAX_VALUE;
    metrics.total_reads = 0;
    metrics.min_writes = Long.MAX_VALUE;
    metrics.total_writes = 0;
    synchronized (this) {
      for (int pid = 0; pid < args.partitions; pid++) {
        var partition = partitions.get(pid);

        metrics.total_reads += partition.countRead;
        metrics.min_reads = Math.min(metrics.min_reads, partition.countRead);
        metrics.total_writes += partition.countWritten;
        metrics.min_writes
            = Math.min(metrics.min_writes, partition.countWritten);

        var info = new App.PartitionMetrics();
        info.partition = partition.id;
        info.end_offset = partition.endOffset;
        info.read_offset = partition.readOffset;
        info.read_position = partition.readPosition;
        info.consumed = partition.consumed;
        info.written_offset = partition.writtenOffset;

        metrics.partitions.add(info);
      }
    }
    return metrics;
  }

  private void readProcess(int pid) throws Exception {
    var partition = partitions.get(pid);

    var tp = new TopicPartition(args.topic, pid);
    var tps = Collections.singletonList(tp);

    Properties props = new Properties();
    props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, args.brokers);
    props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false);
    props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
    props.put(ConsumerConfig.ISOLATION_LEVEL_CONFIG, "read_committed");
    props.put(
        ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG,
        "org.apache.kafka.common.serialization.StringDeserializer");
    props.put(
        ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
        "org.apache.kafka.common.serialization.StringDeserializer");

    HashMap<String, Value> kv = new HashMap<>();

    KafkaConsumer<String, String> consumer = new KafkaConsumer<>(props);
    consumer.assign(tps);
    consumer.seekToEnd(tps);
    long end = consumer.position(tp);
    long written = -1;
    synchronized (this) {
      partition.endOffset = end;
      written = partition.writtenOffset;
    }
    consumer.seekToBeginning(tps);

    long lastOffset = -1;
    long lastOpId = -1;

    while (consumer.position(tp) < end && consumer.position(tp) <= written) {
      synchronized (this) { partition.readPosition = consumer.position(tp); }
      ConsumerRecords<String, String> records
          = consumer.poll(Duration.ofMillis(10000));
      var it = records.iterator();
      int read = 0;
      while (it.hasNext()) {
        read += 1;
        var record = it.next();

        var offset = record.offset();
        if (offset <= lastOffset) {
          violation(pid, "read offset " + offset + " after " + lastOffset);
        }
        lastOffset = offset;
        synchronized (this) { partition.readOffset = lastOffset; }
        var parts = record.value().split("\t");
        long opId = Long.parseLong(parts[0]);
        log(pid, "r\t" + opId + "\t" + offset);
        if (opId <= lastOpId) {
          violation(pid, "read opId " + opId + " after " + lastOpId);
        }
        lastOpId = opId;
        if (!parts[1].equals("c")) {
          violation(
              pid, "read aborted " + record.key() + "=" + opId + "@" + offset);
        }
        kv.put(record.key(), new Value(opId, offset));
      }
      synchronized (this) { partition.countRead += read; }
    }
    synchronized (this) { partition.readPosition = consumer.position(tp); }
    consumer.close();

    var writtenKV = partition.latestKV;

    for (var key : kv.keySet()) {
      if (!writtenKV.containsKey(key)) {
        violation(pid, "read key " + key + " wasn't written");
      }
      var readValue = kv.get(key);
      var writtenValue = writtenKV.get(key);
      if (writtenValue.confirmed != null
          && writtenValue.confirmed.id == readValue.id) {
        if (writtenValue.confirmed.offset != readValue.offset) {
          violation(
              pid, "last observed offset=" + readValue.offset + " for key="
                       + key + " doesn't match last written offset="
                       + writtenValue.confirmed.offset
                       + " for the same value=" + readValue.id);
        }
      } else {
        WriteInfo match = null;
        for (var attempt : writtenValue.attempts) {
          if (attempt.id == readValue.id) {
            match = attempt;
          }
        }
        if (match == null) {
          violation(
              pid, "can't find an attempt matching read value" + readValue.id);
        }
      }
      writtenKV.remove(key);
    }

    for (var key : writtenKV.keySet()) {
      var writtenValue = writtenKV.get(key);
      if (writtenValue.confirmed != null) {
        violation(
            pid, "confirmed write " + key + "=" + writtenValue.confirmed.id
                     + "@" + writtenValue.confirmed.offset
                     + " can't be found in final read");
      }
    }

    synchronized (this) { partition.consumed = true; }

    log(pid, "partition is validated");
  }
}

File: tests/rptest/transactions/verifiers/txworkload.py
#!/usr/bin/env python3

# make sure to `pip install flask confluent_kafka`

import os
import random
import sys
import time
import uuid

from enum import Enum, auto
from threading import Thread, RLock, local as tlocal
from typing import Optional
from concurrent.futures import Future, ThreadPoolExecutor

from confluent_kafka import Producer, Consumer, TopicPartition
from flask import Flask, abort, jsonify, request


class State(Enum):
    INITIAL = auto()
    PRODUCING = auto()
    DRAINING = auto()
    DRAINED = auto()
    CONSUMING = auto()
    FINAL = auto()


class WriteInfo:
    id: int
    key: str
    partition: int
    offset: int

    def __init__(self, id, key, partition, offset):
        self.id = id
        self.key = key
        self.partition = partition
        self.offset = offset


class Value:
    id: int
    offset: int

    def __init__(self, id, offset):
        self.id = id
        self.offset = offset


class LatestValue:
    confirmed: Optional[WriteInfo]
    attempts: list[WriteInfo]

    def __init__(self):
        self.confirmed = None
        self.attempts = []


class GatedWorkload:
    LETTERS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

    _count: int = 0
    _state: State = State.INITIAL

    _started: float
    _write_threads: list[Future[None]]
    _read_threads: list[Future[None]]
    _locks: dict[str, RLock] = {
        'count': RLock(),
        'start_producer': RLock(),
        'stop_producer': RLock(),
        'wait_producer': RLock(),
        'start_consumer': RLock(),
        'wait_consumer': RLock(),
        'write_process': RLock(),
        'read_process': RLock()
    }

    def random_string(self, length: int):
        return ''.join(random.choice(self.LETTERS) for _ in range(length))

    def log(self, partition: int, msg: str):
        ts = (time.time() - self._started) / 1000
        print(f'{self._count}\t{ts}\t{partition}\t{msg}\n')
        with self._locks['count']:
            self._count += 1

    def violation(self, partition: int, msg: str):
        self.log(partition, f'violation\t{msg}')
        sys.exit(1)

    def get_metrics(self):
        raise NotImplementedError

    def start_producer(self):
        with self._locks['start_producer']:
            if self._state != State.INITIAL:
                self.violation(
                    -1, 'producer can be started only from the initial state')
            self._state = State.PRODUCING
        self._started = time.time()

    def stop_producer(self):
        with self._locks['stop_producer']:
            if self._state != State.PRODUCING:
                self.violation(
                    -1,
                    'producer can be stopped only from the producing state')
            self._state = State.DRAINING

    def wait_producer(self):
        with self._locks['wait_producer']:
            match self._state:
                case State.DRAINED:
                    return
                case State.DRAINING:
                    [f.result() for f in self._write_threads]
                    self.log(-1, 'writers stopped')
                    self._state = State.DRAINED
                case _:
                    self.violation(-1, "producer isn't active")

    def start_consumer(self):
        with self._locks['start_consumer']:
            if self._state != State.DRAINED:
                self.violation(
                    -1, 'consumer can be started only from the drained state')
            self._state = State.CONSUMING

    def wait_consumer(self):
        with self._locks['wait_consumer']:
            match self._state:
                case State.FINAL:
                    return
                case State.CONSUMING:
                    [f.result() for f in self._read_threads]
                    self.log(-1, 'readers stopped')
                    self._state = State.FINAL
                case _:
                    self.violation(-1,
                                   "can't wait consumer before starting it")


class TxWorkload(GatedWorkload):
    def __init__(self, args):
        super().__init__()
        self.args = {
            'threads': 4,
            'batch_size': 10
        } | args
        self.partitions = {}
        self.last_offset = -1

    def init_partitions(self):
        for pid in range(self.args['partitions']):
            partition = TopicPartition(f'topic-{pid}', pid)
            self.partitions[pid] = partition

    def start_producer(self):
        super().start_producer()
        print(
            '#logical_time\tseconds_since_start\tpartition\tlog_type\tdetails\n'
        )
        self.log(-1, 'start\tproducer')
        self._write_threads = []
        self.init_partitions()

        def try_write_process(id):
            try:
                self.write_process(id)
            except:
                with self._locks['start_producer']:
                    print('=== write process error\n')
                    print(repr(sys.exception()))
                    os._exit(1)

        with ThreadPoolExecutor(max_workers=self.args['threads']) as exe:
            for partition in range(self.args['partitions']):
                self._write_threads.append(
                    exe.submit(try_write_process, partition))

    def start_consumer(self):
        super().start_consumer()
        self._read_threads = []

        def try_read_process(id):
            try:
                self.read_process(id)
            except:
                with self._locks['start_consumer']:
                    print('=== read process error\n')
                    print(repr(sys.exception()))
                    os._exit(1)

        with ThreadPoolExecutor(max_workers=self.args['threads']) as exe:
            for partition in range(self.args['partitions']):
                self._read_threads.append(
                    exe.submit(try_read_process, partition))

    def write_process(self, pid):
        # Create a dictionary to hold the producer configuration
        config = tlocal()
        config.__dict__.update({
            'bootstrap.servers': self.args['brokers'],
            'enable.idempotence': True,
            'linger.ms': 0,
            'transactional.id': str(uuid.uuid4())
        })

        batch_size = self.args['batch_size']
        abort_probability = self.args['abort_probability']

        id = 0
        last_offset = -1
        should_reset = True
        producer: Optional[Producer] = None

        while self._state == State.PRODUCING:
            if should_reset:
                if producer is not None:
                    producer.close()
                producer = None
                should_reset = False
                time.sleep(1.0)

            if producer is None:
                # Create a Kafka producer
                producer = Producer(config.__dict__)
                try:
                    self.log(pid, 'init')
                    producer.init_transactions()
                except:
                    with self._locks['write_process']:
                        self.log(pid, 'err\tinit')
                        print('=== Error on initTransactions\n')
                        print(repr(sys.exception()))
                    should_reset = True
                    continue

            producer.begin_transaction()
            should_abort = random.random() < abort_probability
            batch: dict[str, WriteInfo] = {}

            def cb(err, msg):
                if err is not None:
                    self.log(-1, f'produce delivery error: {err}')
                    os._exit(1)
                if msg.error() is not None:
                    self.log(-1, f'produce delivery msg error: {msg.error()}')
                    os._exit(1)
                with self._locks['write_process']:
                    key = msg.key()
                    partition = msg.partition()
                    last_offset = batch[key].offset = msg.offset()
                    self.log(partition, f'last_offset\t{last_offset}')

            for _ in range(batch_size):
                op = WriteInfo(id, f'key{id}', pid, -1)
                batch[op.key] = op
                if should_abort:
                    self.log(pid, f"a\t{op.id}")
                else:
                    self.log(pid, f"w\t{op.id}")

                value = f'{op.id}\t{"a" if should_abort else "c"}\t{self.random_string(1024)}'

                producer.produce(self.args['topic'], value, op.key, pid, cb)
                # this doesn't work - producer.produce returns None
                with self._locks['write_process']:
                    id += 1

            if should_abort:
                producer.flush()
                producer.abort_transaction()
            else:
                producer.commit_transaction()

    def read_process(self, pid):
        config = tlocal()
        config.__dict__.update({
            'bootstrap.servers': self.args['brokers'],
            'enable.auto.commit': False,
            'isolation.level': 'read_committed',
            'auto.offset.reset': 'earliest',
            'group.id': 'my-group'
        })
        consumer = Consumer(config.__dict__)
        partition = self.partitions[pid]
        # TODO: finish porting this over from Java

    def get_metrics(self):
        # TODO: implement metrics
        return {}


def main():
    app = Flask(__name__)
    workload: Optional[GatedWorkload] = None

    @app.route('/ping', methods=['GET'])
    def ping():
        return ''

    @app.route('/info', methods=['GET'])
    def info():
        if workload is None:
            abort(409)
        return workload.get_metrics()

    # curl -vX POST http://127.0.0.1:8080/start-producer -H 'Content-Type:
    # application/json' -d '{"abort_probability":0.1,"batch_size":10,
    # "brokers":"127.0.0.1:9092","partitions":1,"workload":"TX"}'
    @app.route('/start-producer', methods=['POST'])
    def start_producer():
        params = request.get_json()
        match params['workload']:
            case 'TX':
                workload = TxWorkload(params)
            case _:
                print(f'unknown workload: "{params["workload"]}"')
                raise NotImplementedError
        workload.start_producer()
        return ''

    @app.route('/stop-producer', methods=['POST'])
    def stop_producer():
        if workload is None:
            abort(409)
        workload.stop_producer()
        return ''

    @app.route('/wait-producer', methods=['POST'])
    def wait_producer():
        if workload is None:
            abort(409)
        workload.wait_producer()
        return ''

    @app.route('/start-consumer', methods=['POST'])
    def start_consumer():
        if workload is None:
            abort(409)
        workload.start_consumer()
        return ''

    @app.route('/wait-consumer', methods=['POST'])
    def wait_consumer():
        if workload is None:
            abort(409)
        workload.wait_consumer()
        return ''

    def run_flask(*args):
        port = int(os.getenv('PORT', '8080'))
        app.run(port=port, debug=False)

    main_thread = Thread(target=run_flask)
    main_thread.start()
    main_thread.join()


if __name__ == '__main__':
    main()
